src/level/level_loader.cpp

```diff
@@ -0,0 +1,1201 @@
+        { "CreateFog",          MakeCommand(&CLevelLoader::HandleCommand_CreateFog) },
+        { "CreateLight",        MakeCommand(&CLevelLoader::HandleCommand_CreateLight) },
+        { "CreateSpot",         MakeCommand(&CLevelLoader::HandleCommand_CreateSpot) },
+        { "GroundSpot",         MakeCommand(&CLevelLoader::HandleCommand_GroundSpot) },
+        { "WaterColor",         MakeCommand(&CLevelLoader::HandleCommand_WaterColor) },
+        { "MapColor",           MakeCommand(&CLevelLoader::HandleCommand_MapColor) },
+        { "MapZoom",            MakeCommand(&CLevelLoader::HandleCommand_MapZoom) },
+        { "MaxFlyingHeight",    MakeCommand(&CLevelLoader::HandleCommand_MaxFlyingHeight) },
+        { "AddFlyingHeight",    MakeCommand(&CLevelLoader::HandleCommand_AddFlyingHeight) },
+        { "Camera",             MakeCommand(&CLevelLoader::HandleCommand_Camera, ALL_LOAD_LEVEL_MODES) },
+        { "EndMissionTake",     MakeCommand(&CLevelLoader::HandleCommand_EndMissionTake) },
+        { "EndMissionDelay",    MakeCommand(&CLevelLoader::HandleCommand_EndMissionDelay) },
+        { "EndMissionResearch", MakeCommand(&CLevelLoader::HandleCommand_EndMissionResearch) },
+        { "ObligatoryToken",    MakeCommand(&CLevelLoader::HandleCommand_ObligatoryToken) },
+        { "ProhibitedToken",    MakeCommand(&CLevelLoader::HandleCommand_ProhibitedToken) },
+        { "EnableBuild",        MakeCommand(&CLevelLoader::HandleCommand_EnableBuild) },
+        { "EnableResearch",     MakeCommand(&CLevelLoader::HandleCommand_EnableResearch) },
+        { "DoneResearch",       MakeCommand(&CLevelLoader::HandleCommand_DoneResearch, LevelLoadMode::Normal) },
+        { "NewScript",          MakeCommand(&CLevelLoader::HandleCommand_NewScript) }
+};
+
+CLevelLoader::CLevelLoader(bool solutionEnabled)
+{
+    m_main = CRobotMain::GetInstancePointer();
+    m_solutionEnabled = solutionEnabled;
+    m_nextObjectId = 0;
+    m_objectsNumber = -1;
+    m_backgroundPath = "";
+    m_backgroundUp = Gfx::Color(0.0f, 0.0f, 0.0f, 0.0f);
+    m_backgroundDown = Gfx::Color(0.0f, 0.0f, 0.0f, 0.0f);
+    m_backgroundCloudUp = Gfx::Color(0.0f, 0.0f, 0.0f, 0.0f);
+    m_backgroundCloudDown = Gfx::Color(0.0f, 0.0f, 0.0f, 0.0f);
+    m_backgroundFull = false;
+    m_selectedObject = nullptr;
+}
+
+CLevelLoader::Command::Command(CommandFunction function, int executionConditions)
+{
+    this->function = function;
+    this->executionConditions = executionConditions;
+}
+
+void CLevelLoader::CreateScene(bool solutionEnabled, bool fixScene, LevelLoadMode mode)
+{
+    std::unique_ptr<CLevelLoader> levelLoader = MakeUnique<CLevelLoader>(solutionEnabled);
+
+    levelLoader->ResetScene(mode);
+
+    levelLoader->BuildScene(fixScene, mode);
+}
+
+void CLevelLoader::ResetScene(LevelLoadMode mode)
+{
+    m_main->m_base = nullptr;
+
+    if (mode != LevelLoadMode::Reset)
+    {
```

Comment:
I think the level loader is not really a good place for this function. It's related to resetting the engine state, not actually loading anything.